SeQuant/core/space.hpp

```diff
@@ -417,109 +314,446 @@ class IndexSpace {
   }
   IndexSpace(IndexSpace &&other) {
     if (!other.attr().is_valid())
       throw std::invalid_argument(
           "IndexSpace move ctor received invalid argument");
-    attr_ = other.attr_;
+    attr_ = other.get_attr();
+base_key_ = other.get_base_key();
+approximate_size_ = other.get_approximate_size();
   }
   IndexSpace &operator=(const IndexSpace &other) {
     if (!other.attr().is_valid())
       throw std::invalid_argument(
           "IndexSpace copy assignment operator received invalid argument");
-    attr_ = other.attr_;
+    attr_ = other.get_attr();
+    base_key_ = other.get_base_key();
+    approximate_size_ = other.get_approximate_size();
     return *this;
   }
   IndexSpace &operator=(IndexSpace &&other) {
     if (!other.attr().is_valid())
       throw std::invalid_argument(
           "IndexSpace move assignment operator received invalid argument");
-    attr_ = other.attr_;
+    attr_ = other.get_attr();
+    base_key_ = other.get_base_key();
+    approximate_size_ = other.get_approximate_size();
     return *this;
   }
 
-  /// @brief constructs an instance of an IndexSpace object
-  IndexSpace(Type type, QuantumNumbers qns) noexcept : attr_(type, qns) {
-    assert(attr_.is_valid());
+
+  const Attr get_attr()const {
+    return attr_;
+  }
+  std::wstring get_base_key() const{
+    return base_key_;
+  }
+  static std::wstring_view reduce_key(std::wstring_view key) {
+    const auto underscore_position = key.rfind(L'_');
+    if(underscore_position != -1){ // key can be reduced
+      return key.substr(0, underscore_position);
+    }
+    else{
+      return key;
+    }
+  }
+
+  // having approximate sizes of spaces can be helpful when optimizing evaluation order.
+  unsigned long get_approximate_size() const {
+    return approximate_size_;
   }
 
  private:
   Attr attr_ = Attr::invalid();
+  std::wstring base_key_;
+  unsigned long approximate_size_;
   /// @brief constructs an instance of an IndexSpace object
-  explicit IndexSpace(Attr attr) noexcept : attr_(attr) {
-    assert(attr.is_valid());
+  //explicit IndexSpace(Attr attr) noexcept : attr_(attr) {
+  //  assert(attr.is_valid());
+  //}
+
+
+
+  static std::wstring to_wstring(std::wstring_view key) {
+    return std::wstring(key.begin(), key.end());
   }
 
-  static container::map<Attr, std::wstring> attr2basekey_;
-  static container::map<std::wstring, Attr, KeyCompare> key2attr_;
-  static container::map<Attr, IndexSpace> instances_;
-  static IndexSpace null_instance_;
+};
 
-  static std::wstring_view reduce_key(std::wstring_view key) {
-    const auto underscore_position = key.rfind(L'_');
-    return key.substr(0, underscore_position);
+// An IndexSpaceRegistry is a context specific, modifiable, list of registered indices.
+// It is the task of the USER to ensure that all spaces for their context are included.
+// the result of certain set operations such as intersection and unIon require access to the registry
+//
+class IndexSpaceRegistry{
```

Comment:
I would propose putting this into its own header file for better discoverability